### PR TITLE
machined userns fixes

### LINKED ISSUE #4078: machined: "machinectl copy-from" and "machinectl copy-to" need to adjust uid/gid accordingl to userns if containers uses them
Subject says it all...

See: https://lists.freedesktop.org/archives/systemd-devel/2016-September/037418.html

### PR DESCRIPTION
**Pull request by poettering**
Mostly some infrastructure to fix #4078, and eventually the fix for it.

### CODE REVIEW

File: src/machine/machine-dbus.c [hunk 1]
```diff
@@ -1293,6 +1319,7 @@ const sd_bus_vtable machine_vtable[] = {
         SD_BUS_METHOD("Kill", "si", NULL, bus_machine_method_kill, SD_BUS_VTABLE_UNPRIVILEGED),
         SD_BUS_METHOD("GetAddresses", NULL, "a(iay)", bus_machine_method_get_addresses, SD_BUS_VTABLE_UNPRIVILEGED),
         SD_BUS_METHOD("GetOSRelease", NULL, "a{ss}", bus_machine_method_get_os_release, SD_BUS_VTABLE_UNPRIVILEGED),
+        SD_BUS_METHOD("GetUIDGIDBase", NULL, "u", bus_machine_method_get_uid_gid_base, SD_BUS_VTABLE_UNPRIVILEGED),
```
Comment (martinpitt) [inline src/machine/machine-dbus.c]: Nitpick: as this doesn't have separate out parameters for uid and gid and is hard to read that way, WDYT about "UGID" instead? Also, Base → Offset or Shift?

File: src/machine/machine.c [hunk 2]
```diff
@@ -604,6 +605,100 @@ void machine_release_unit(Machine *m) {
...
+        errno = 0;
+        k = fscanf(f, UID_FMT " " UID_FMT " " UID_FMT, &uid_base, &uid_shift, &uid_range);
+        if (k != 3) {
+                if (ferror(f))
+                        return -errno;
+
+                return -EBADMSG;
+        }
+
+        /* Not a mapping statring at 0? Then it's a complex mapping we can't expose here. */
```
Comment (martinpitt) [inline src/machine/machine.c]: "statring" typo

File: src/machine/machine.c [hunk 3]
```diff
@@ -604,6 +605,100 @@ void machine_release_unit(Machine *m) {
...
+                return -ENXIO;
+        /* Insist that at least the nobody user is mapped, everything else is weird, and hence complex, and we don't support it */
+        if (uid_range < (uid_t) 65534U)
+                return -ENXIO;
+
+        /* If there's more than one line, then we don't support this mapping. */
+        if (fgetc(f) != '\n')
+                return -ENXIO;
+        if (fgetc(f) != EOF)
+                return -ENXIO;
```
Comment (martinpitt) [inline src/machine/machine.c]: The first fgetc() asserts that the first line is always \n terminated. If we can assume that, wouldn't it be easier to read it in the fscanf() above already? If we don't want to assume that, maybe the first fgetc() should check for '\n' or EOF?

File: src/machine/machine-dbus.c [hunk 4]
```diff
@@ -1097,6 +1099,14 @@ int bus_machine_method_copy(sd_bus_message *message, void *userdata, sd_bus_erro
         if (r == 0)
                 return 1; /* Will call us back */
 
+        r = machine_get_uid_gid_base(m, &uid);
+        if (r < 0)
+                return r;
+
+        /* Is user namespacing is used? ifso set the resulting UIDs always to zero */
+        if (uid != 0)
+                copy_flags |= COPY_CHOWN_ROOT;
```
Comment (martinpitt) [inline src/machine/machine-dbus.c]: AFAICS this is run in machined, i. e. the host's user namespace, not the container's.  How does this make sure that "root" (uid 0) of the container is being used here, i. e. the shifted UID?

**Comment by evverx** [comment]
> Mostly some infrastructure to fix #4078, and eventually the fix for it

Well, I'm reading https://lists.freedesktop.org/archives/systemd-devel/2016-September/037421.html
> This issue is new and have been able to cp/mv from host to container and
preserve file/folders attributes until now. Something in my recent upgrades
have done these changes.

I think the original issue unrelated to `machinectl [copy-from|copy-to]`. Arnaud Gaboury tried to copy/move some files manually. And this doesn't work.

I  guess the docs should contain something like:
```
you shouldn't even try to cp/mv files from/to `-U`-container manually
```

Also, I'm not sure how does this PR address `bind`: https://github.com/systemd/systemd/issues/4078#issuecomment-268043354

File: src/machine/machine-dbus.c (hunk 1 above)
Comment (poettering) [reply src/machine/machine-dbus.c]: Hmm, internally we actually do call this "shift", hence I figure you are right, we should expose that as shift too. Will change.

I figure UID is the relevant bit here, and GIDs ranges are just always kept in sync with UIDs, and generally secondary, hence I figure we should drop the explicit reference to GIDs entirely.

I'll hence change this to become GetUIDShift() now, I hope that makes sense!

File: src/machine/machine-dbus.c (hunk 4 above)
Comment (poettering) [reply src/machine/machine-dbus.c]: uh, it didn't! I only tested copy from the container, not copy to the container, hence I didn't notice. I reworked this now, to fix it up in both directions!

File: src/machine/machine.c (hunk 3 above)
Comment (poettering) [reply src/machine/machine.c]: true, i added the \n now to the fscanf() format string!

**Comment by poettering** [comment]
I have now force pushed a new, rebased version that addresses all the issues raised.

@evverx i also added code to effectively block "machinectl bind" if userns is used, returning a brief explanatory error message. It's the best we can do until the kernel gains a "shiftfs" or similar concept.

I also added the suggested man page addition saying that copying files from/to the container image directly requires patching of UIDs/GIDs.

File: src/machine/machine-dbus.c [hunk 5]
```diff
@@ -875,6 +876,12 @@ int bus_machine_method_bind_mount(sd_bus_message *message, void *userdata, sd_bu
         if (r == 0)
                 return 1; /* Will call us back */
 
+        r = machine_get_uid_shift(m, &uid);
+        if (r < 0)
+                return r;
+        if (uid != 0)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Can't bind mount on container with user namespacing applied.");
```
Comment (martinpitt) [inline src/machine/machine-dbus.c:883]: Hmm, the most likely result of bind-mounting with shifted UIDs is that the container can only read the files, but not  write anything into it; this is certainly a common use case, and not particularly unsafe?

File: man/systemd-nspawn.xml [hunk 6]
```diff
@@ -474,7 +474,12 @@
 
         <para>Note that the picked UID/GID range is not written to <filename>/etc/passwd</filename> or
         <filename>/etc/group</filename>. In fact, the allocation of the range is not stored persistently anywhere,
-        except in the file ownership of the files and directories of the container.</para></listitem>
+        except in the file ownership of the files and directories of the container.</para>
+
+        <para>Note that when user namespacing is used file ownership on disk reflects this, and all of the container's
+        files and directories are owned by the container's effective user and group IDs. This means that copying files
+        from and to the container image requires correction of the numeric UID/GID values, according to the UID/GID
+        shift applied.</para></listitem>
```
Comment (martinpitt) [inline man/systemd-nspawn.xml:482]: Could you please change the commit message to use less strong words?

File: src/machine/machine-dbus.c (hunk 5 above)
Comment (poettering) [reply src/machine/machine-dbus.c:883]: Well, information leakage is also a problem. It's not sufficient to say that we prohibit write access, it's also not safe giving read access to files that a process should not have read access to. Worse even: if the bind mount contains suid files, then the mismatch might possibly even permit somebody who shouldn't have privs get privs. I'd rather not get into that and I think it's simpler to simply step away from it... In the long run I hope that shiftfs in the kernel will fix this properly for us.

File: man/systemd-nspawn.xml (hunk 6 above)
Comment (poettering) [reply man/systemd-nspawn.xml:482]: done!

File: src/machine/machine-dbus.c (hunk 5 above)
Comment (martinpitt) [reply src/machine/machine-dbus.c:883]: OK, good point. I have no strong opinion about it, so let's land this and revisit if someone actually complains?

**Pull request merged**